src/main/java/com/google/firebase/database/connection/NettyWebSocketClient.java

```diff
@@ -0,0 +1,197 @@
+
+import java.net.URI;
+import java.security.KeyStore;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadFactory;
+import javax.net.ssl.TrustManagerFactory;
+
+/**
+ * A {@link WebsocketConnection.WSClient} implementation based on the Netty framework. Uses
+ * a single-threaded NIO event loop to read and write bytes from a WebSocket connection. Netty
+ * handles all the low-level IO, SSL and WebSocket handshake, and other protocol-specific details.
+ *
+ * <p>This implementation does not initiate connection close on its own. In case of errors or loss
+ * of connectivity, it notifies the higher layer ({@link WebsocketConnection}), which then decides
+ * whether to initiate a connection tear down.
+ */
+class NettyWebSocketClient implements WebsocketConnection.WSClient {
+
+  private static final int DEFAULT_WSS_PORT = 443;
+
+  private final URI uri;
+  private final WebsocketConnection.WSClientEventHandler eventHandler;
+  private final ChannelHandler channelHandler;
+  private final ExecutorService executorService;
+  private final EventLoopGroup group;
+
+  private Channel channel;
+
+  NettyWebSocketClient(
+      URI uri, String userAgent, ThreadFactory threadFactory,
+      WebsocketConnection.WSClientEventHandler eventHandler) {
+    this.uri = checkNotNull(uri, "uri must not be null");
+    this.eventHandler = checkNotNull(eventHandler, "event handler must not be null");
+    this.channelHandler = new WebSocketClientHandler(uri, userAgent, eventHandler);
+    this.executorService = new RevivingScheduledExecutor(
+        threadFactory, "firebase-websocket-worker", GaeThreadFactory.isAvailable());
+    this.group = new NioEventLoopGroup(1, this.executorService);
+  }
+
+  @Override
+  public void connect() {
+    checkState(channel == null, "channel already initialized");
+    try {
+      TrustManagerFactory trustFactory = TrustManagerFactory.getInstance(
+          TrustManagerFactory.getDefaultAlgorithm());
+      trustFactory.init((KeyStore) null);
+      final SslContext sslContext = SslContextBuilder.forClient()
+          .trustManager(trustFactory).build();
+      Bootstrap bootstrap = new Bootstrap();
+      final int port = uri.getPort() != -1 ? uri.getPort() : DEFAULT_WSS_PORT;
+      bootstrap.group(group)
+          .channel(NioSocketChannel.class)
+          .handler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            protected void initChannel(SocketChannel ch) {
+              ChannelPipeline p = ch.pipeline();
+              p.addLast(sslContext.newHandler(ch.alloc(), uri.getHost(), port));
+              p.addLast(
+                  new HttpClientCodec(),
+                  // Set the max size for the HTTP responses. This only applies to the WebSocket
+                  // handshake response from the server.
+                  new HttpObjectAggregator(32 * 1024),
+                  WebSocketClientCompressionHandler.INSTANCE,
```

Comment:
We don't compress websocket frames and don't plan on it, so this probably isn't needed

Reply:
Done. Thanks @rockwotj